### PR TITLE
[RBAC] BREAKING CHANGE: Fix #11883: `az role assignment create`: empty scope will prompt error

### LINKED ISSUE #11883: az role assignment create scope default to subscription if empty
> ### `az feedback` auto-generates most of the information requested below, as of CLI version 2.0.62

**Describe the bug**
# Summary

Using `azure-cli` (in actual version of `2.0.79`) and `az role assignment create` command.

The problem I see is the use of `--scope` parameter.
Empty (I mean empty = specified but undefined parameter) scope equals to default (therefore subscription) scope.

Why is that? How about the error message for this action? User is not aware of the issue and scope automatically applies on the subscription.
That can be harmful in some cases (see examples below) - user or service principal can be assigned more powerfull scope and you can imagine the consequences.

# Description

Following examples can cause elevation of privileges. In case of resource name typo, non-existing resource, etc...

## Steps to Reproduce:

```bash
# login to the portal
az login
# obtain assignee identifier (user in this case)
ASSIGNEE=$(az account show --query 'user.name' -o tsv)

# acceptable version: undefined scope => default scope => applied to subscription, OK
az role assignment create \
--assignee "${ASSIGNEE}" \
--role Contributor

# Why empty (empty, not undefined!) scope equals to default scope? Following examples can cause elevation of privileges.

# version 1: EMPTY scope => applied to subscription (nonintended elevation of privileges)
az role assignment create \
--assignee "${ASSIGNEE}" \
--scope "" \
--role Contributor

# version 2: non existing resource => EMPTY scope => applied to subscription (nonintended elevation of privileges)
az role assignment create \
--assignee "${ASSIGNEE}" \
--scope "$(az storage account show --name iMakeATypoInName --subscription ${AZURE_SUBSCRIPTION} --query 'id' -o tsv)" \
--role Contributor

# version 3: non existing resource in different way => EMPTY scope => applied to subscription (nonintended elevation of privileges)
OBJECT=$(az storage account show --name iMakeATypoInName --subscription ${AZURE_SUBSCRIPTION} --query 'id' -o tsv)
az role assignment create \
--assignee "${ASSIGNEE}" \
--scope "${OBJECT}" \
--role Contributor

# check for all the three versions:
az role assignment list --assignee "${ASSIGNEE}"
```

```bash
# output:
[
{
"canDelegate": null,
"id": "/subscriptions/<subscription_guid>/providers/Microsoft.Authorization/roleAssignments/<assignment_guid>",
"name": "<assignment_guid>",
"principalId": "<principal_id>",
"principalName": "andrej.zelnik@....",
"principalType": "User",
"roleDefinitionId": "/subscriptions/<subscription_guid>/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c",
"roleDefinitionName": "Contributor",
"scope": "/subscriptions/<subscription_guid>",
"type": "Microsoft.Authorization/roleAssignments"
}
]

```

## Supporting materials/ references:

* https://docs.microsoft.com/en-us/cli/azure/role/assignment?view=azure-cli-latest#az-role-assignment-create

**To Reproduce**
Steps to reproduce the behavior.

**Expected behavior**

**Environment summary**
Install Method (e.g. pip, interactive script, apt-get, Docker, MSI, edge build) / CLI version (`az --version`) / OS version / Shell Type (e.g. bash, cmd.exe, Bash on Windows)  

**Additional context**

### PR DESCRIPTION
**Pull request by arrownj**
This is a PR to fix issue #11883 

The previous design can cause elevation of privileges. 

In the original PR #11977 , we plan to fix this problem by adding a validation for `--scope`, the validator will promt error when `--scope` is not specified or specified an empty string. And the change will impact `assignment create`, `assignment delete` and `assignment list`. Of cause, this is a breaking change to end users.

As discussed in https://github.com/Azure/azure-cli/pull/11977#issuecomment-579774320, we decided not to change the behavior when `--scope` is not specified.

Just as user said in origin issue(paste below), when `--scope` is not specified, it is OK to use default subscription value. What user concerns is that when `--scope` is specified an empty value, it shouldn't use subscription anymore.

```bash
# login to the portal
az login
# obtain assignee identifier (user in this case)
ASSIGNEE=$(az account show --query 'user.name' -o tsv)

# acceptable version: undefined scope => default scope => applied to subscription, OK
az role assignment create \
--assignee "${ASSIGNEE}" \
--role Contributor

# Why empty (empty, not undefined!) scope equals to default scope? Following examples can cause elevation of privileges.

# version 1: EMPTY scope => applied to subscription (nonintended elevation of privileges)
az role assignment create \
--assignee "${ASSIGNEE}" \
--scope "" \
--role Contributor
```

So we finally change the solution to promt error only when `--scope` is specified an empty value. This is still a breaking chang(to user who specify an empty value to `--scope` and expect subscription  value), however, the impacts are much reduced compared to previous solution.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S165.

File: src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py [hunk 1]
```diff
@@ -295,6 +296,11 @@ def test_role_assignment_e2e(self, resource_group):
                 self.cmd('role assignment list --assignee {upn}',
                          checks=self.check("length([])", 1))
                 self.cmd('role assignment delete --assignee {upn} --role reader')
+
+                # test role assignment on empty scope
+                with self.assertRaisesRegexp(CLIError, 'Invalid scope. Please use --help to view the valid format.'):
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py]: Usually we just use some phrases in the error message, like "Invalid scope", to have more flexibility of the error message in the future. But this is totally fine to be more strict on the error message. 😉

File: src/azure-cli/HISTORY.rst [hunk 2]
```diff
@@ -3,6 +3,10 @@
 Release History
 ===============
 
+**RBAC**
+
+* [BREAKING CHANGE] Fix #11883: `az role assignment create`: empty scope will prompt error
```
Comment (jiasli) [inline src/azure-cli/HISTORY.rst]: Please place this entry in the right place in the history file following alphabetic order.

Comment (arrownj) [reply src/azure-cli/HISTORY.rst]: Thanks, updated.

**Comment by jiasli** [comment]
It seems we missed another case where `--resource-group ""`, this will again fallback to subscription level assignment, with even incorrect `principalType`:

```
az role assignment create --assignee-object-id 0d504196-1423-4569-9a6e-15149656f0ee --role "Contributor" --resource-group ""
{
  "canDelegate": null,
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleAssignments/4290ea7b-9cca-414f-9f0d-eeb931cc9e40",
  "name": "4290ea7b-9cca-414f-9f0d-eeb931cc9e40",
  "principalId": "00215dd3-8494-4737-8f44-47c950eee2a8",
  "principalName": "",
  "principalType": "ServicePrincipal",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c",
  "roleDefinitionName": "Contributor",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590",
  "type": "Microsoft.Authorization/roleAssignments"
}
```

**Comment by yungezz** [comment]
let's create a new issue for the resourcegroup empty string. Worth to check policy assignment/definition also.

**Pull request merged**